agent/src/main/java/com/microsoft/applicationinsights/agent/internal/agent/AdvancedAdviceAdapter.java

```diff
@@ -0,0 +1,169 @@
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.agent.internal.agent;
+
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.AdviceAdapter;
+
+/**
+ * A base class that lets its derived classes to enjoy utilities
+ * like creating temporary variables, managing methods points, method names etc.
+ *
+ * Created by gupele on 5/11/2015.
+ */
+abstract class AdvancedAdviceAdapter extends AdviceAdapter {
+
+    protected final boolean reportExecutionTime;
+
+    protected enum ExitStatus {
+        EXIT_WITH_EXCEPTION,
+        EXIT_WITH_RETURN_VALUE,
+        EXIT_VOID,
+        EXIT_UNKNOWN
+    }
+
+    protected static class TempVar {
+        public final int tempVarIndex;
+
+        public TempVar(int tempVarIndex) {
+            this.tempVarIndex = tempVarIndex;
+        }
+    }
+
+    protected static class TempArrayVar {
+        public final int tempVarIndex;
+
+        public TempArrayVar(int tempVarIndex) {
+            this.tempVarIndex = tempVarIndex;
+        }
+    }
+
+    private String methodName;
+    private String desc;
+
+    private Label startTryFinallyBlock = new Label();
+    private Label endTryFinallyBlock = new Label();
+
+    @Override
+    public void visitCode() {
+        super.visitCode();
+        mark(startTryFinallyBlock);
+    }
+
+    @Override
+    public void visitMaxs(int maxStack, int maxLocals) {
+        visitTryCatchBlock(startTryFinallyBlock, endTryFinallyBlock, endTryFinallyBlock, null);
+        mark(endTryFinallyBlock);
+
+        byteCodeForMethodExit(Opcodes.ATHROW);
+
+        mv.visitInsn(Opcodes.ATHROW);
+        mv.visitMaxs(maxStack, maxLocals);
+    }
+
+    @Override
+    protected void onMethodExit(int opcode) {
+        if (!reportExecutionTime) {
+            return;
+        }
+
+        if (opcode != Opcodes.ATHROW) {
+            byteCodeForMethodExit(opcode);
+        }
+    }
+
+    @Override
+    public void visitMethodInsn(int opcode,String owner,String name,String desc, boolean isMethodOwnerAnInterface) {
```

Comment:
Q - if you're just delegating to your super class do you really need to override this method?